### PR TITLE
Training gpt-neox 20B using neuronx-distributed

### PR DESCRIPTION
**Pull request by XinweiFu**
This PR provides the scripts of training gpt-neox 20B using neuronx-distributed.

### CODE REVIEW

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/adamw_fp32_optim_params.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+# coding=utf-8
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/adamw_fp32_optim_params.py:1]: Can we move this file out since it has been shared by alot of models

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/get_dataset.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+from datasets import load_dataset
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/get_dataset.py:1]: Same for this. Either we can have a common folder for both 6.9B and 20B

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+_REAL_CHECKPOINT_FOR_DOC = "EleutherAI/gpt-neox-20b"
+_CONFIG_FOR_DOC = "GPTNeoXConfig"
+
+GPT_NEOX_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "EleutherAI/gpt-neox-20b",
+    # See all GPTNeoX models at https://huggingface.co/models?filter=gpt_neox
+]
+
+
+class GPTNeoXPreTrainedModel(PreTrainedModel):
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:58]: Did we change anything from Transformer's NeoX class?

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/adamw_fp32_optim_params.py (hunk 1 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/adamw_fp32_optim_params.py:1]: Yes. We can remove it and reuse the one in 6.9B.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/get_dataset.py (hunk 2 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/get_dataset.py:1]: Yes. We can remove it and reuse the one in 6.9B.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, GPTNeoXModel):
+            module.gradient_checkpointing = value
+
+
+class GPTNeoXAttention(nn.Module):
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:88]: Are we touching each an every function of the original class. Just checking if we can inherit the original class and override those functions which we need. This way we can show the users what are the main changes and where we deviate. We can add comments too in the places we add new changes

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+        attn_output = torch.matmul(attn_weights, value)
+        return attn_output, attn_weights
+
+
+def attention_mask_func(attention_scores, ltor_mask):
+    attention_scores.masked_fill_(~ltor_mask, torch.finfo(attention_scores.dtype).min)
+    return attention_scores
+
+
+class RotaryEmbedding(torch.nn.Module):
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:284]: What did we change in this class?

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+            t = torch.arange(self.max_seq_len_cached, device=x.device, dtype=self.inv_freq.dtype)
+            freqs = torch.einsum("i,j->ij", t, self.inv_freq)
+            # Different from paper, but it uses a different permutation in order to obtain the same calculation
+            emb = torch.cat((freqs, freqs), dim=-1).to(x.device)
+            self.cos_cached = emb.cos()[None, None, :, :]
+            self.sin_cached = emb.sin()[None, None, :, :]
+        return self.cos_cached[:seq_len, ...].to(x.device), self.sin_cached[:seq_len, ...].to(x.device)
+
+
+def rotate_half(x):
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:313]: Can we reuse from original transformers?

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, offset: int = 0):
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:320]: Can we reuse from original transformers?

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 8]
```diff
@@ -0,0 +1,818 @@
...
+    cos = cos[..., offset : q.shape[-2] + offset, :]
+    sin = sin[..., offset : q.shape[-2] + offset, :]
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+
+class GPTNeoXMLP(nn.Module):
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:328]: We can inherit the original class and re-write just the init?

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 3 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:58]: We changed a lot:
- replace linears in attention and mlp
- replace embedding and lm_head
- use parallel cross entropy
- modifications for sequence parallel and mask creation on-the-fly

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py [hunk 9]
```diff
@@ -0,0 +1,818 @@
+# coding=utf-8
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py:1]: Overall can we avoid copying the entire file and override functions whereever needed? This would make it easier to upstream too and also help users to understand what changes are needed to convert their existing models to use TP+SP

**Comment by XinweiFu** [comment]
@aws-rhsoln I got your point. I will do a revision by creating a model file which only contains the changes from the original model.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/tp_dp_gpt_neox_20b_hf_pretrain.py [hunk 10]
```diff
@@ -0,0 +1,659 @@
+# coding=utf-8
```
Comment (aws-rhsoln) [inline torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/tp_dp_gpt_neox_20b_hf_pretrain.py:1]: Is it possible to re-use the existing training script from GPT-NeoX-6.9B, you can then parameterize it with some config json. If 6.9B config is passed, then it would pick 6.9B mmodle and if 20B is passed then it would pick 20B model. That way 6.9B model can also enjoy the SP+TP changes that you added.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/adamw_fp32_optim_params.py (hunk 1 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/adamw_fp32_optim_params.py]: Removed.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/get_dataset.py (hunk 2 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/get_dataset.py]: Removed.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 3 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 5 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 7 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 6 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 8 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 9 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/tp_dp_gpt_neox_20b_hf_pretrain.py (hunk 10 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/tp_dp_gpt_neox_20b_hf_pretrain.py:1]: Have cleaned up the code by deleting unnecessary codes.

File: torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py (hunk 4 above)
Comment (XinweiFu) [reply torch-neuronx/training/tp_dp_gpt_neox_20b_hf_pretrain/modeling_gpt_neox_nxd.py]: Have cleaned up the code by deleting unnecessary codes.

**Comment by natemail-aws** [comment]
Can we create a jupyter notebook for this sample? Also can we close the PR on the public repo

**Comment by XinweiFu** [comment]
Moved the PR to here: https://github.com/aws-neuron/aws-neuron-samples-staging/pull/131
